osu.Game.Tests/Visual/Gameplay/TestSceneSliderPath.cs

```diff
@@ -185,6 +185,36 @@ public void TestShortenToNegativeLength()
             AddStep("shorten to -10 length", () => path.ExpectedDistance.Value = -10);
         }
 
+        [Test]
+        public void TestGetSegmentEnds()
+        {
+            var positions = new[]
+            {
+                Vector2.Zero,
+                new Vector2(100, 0),
+                new Vector2(100),
+                new Vector2(200, 100),
+            };
+            double[] distances = { 100d, 200d, 300d };
+
+            AddStep("create path", () => path.ControlPoints.AddRange(positions.Select(p => new PathControlPoint(p, PathType.Linear))));
+            AddAssert("segment ends are correct", () => path.GetSegmentEnds(), () => Is.EqualTo(distances.Select(d => d / 300)));
+            AddAssert("segment end positions recovered", () => path.GetSegmentEnds().Select(p => path.PositionAt(p)), () => Is.EqualTo(positions.Skip(1)));
+
+            AddStep("lengthen last segment", () => path.ExpectedDistance.Value = 400);
+            AddAssert("segment ends are correct", () => path.GetSegmentEnds(), () => Is.EqualTo(distances.Select(d => d / 400)));
+            AddAssert("segment end positions recovered", () => path.GetSegmentEnds().Select(p => path.PositionAt(p)), () => Is.EqualTo(positions.Skip(1)));
+
+            AddStep("shorten last segment", () => path.ExpectedDistance.Value = 150);
+            AddAssert("segment ends are correct", () => path.GetSegmentEnds(), () => Is.EqualTo(distances.Select(d => d / 150)));
+            AddAssert("segment end positions not recovered", () => path.GetSegmentEnds().Select(p => path.PositionAt(p)), () => Is.EqualTo(new[]
+            {
+                positions[1],
+                new Vector2(100, 50),
+                new Vector2(100, 50),
+            }));
```

Comment:
To be quite honest, I'm having trouble with these test cases because I'm not sure they make sense to me.

- For the "lengthened path" case, I'd expect the last segment to terminate at 1, not at presumably 0.75 that it's gonna do now.
- For the "shortened path" case, I guess clamping the truncated segments to 1 makes sense, but I'd also accept like a `null` or something.

It very strongly depends what we want out of this method, I guess, and I'm not sure _what_ we want out of it.

- If what we want is literally the path progress value for each last control point of the segment truncated to [0,1] (which is what this is), then this is fine. But I'm not sure I'd call it "segment ends" either.
- If what we want is _actual geometrical_ segment ends, then the lengthen case is definitely wrong (but the shorten case is probably okay).

Not sure. cc @Pasi4K5 since this influences you too.

Reply:
It appears like you don't understand the output of the method.

> If what we want is literally the path progress value for each last control point of the segment truncated to [0,1] (which is what this is)

That is not what this is, because `GetSegmentEnds` does not truncate the values. We want untruncated values.

What I want out of the method is values that you would put into `PositionAt` to get the positions of the segment ends. 
- In the lengthened case, the `Distance` is more than the distance to the last segment end, so the last segment ends at 75% of `Distance`, hence it returns `0.75`. The last 25% is just extention that doesn't belong to any segment.
- In the shortened case, the `Distance` is less than the distance to the last segment end, so the last segment ends at 200% of `Distance`, hence it returns `2`. The only reason that the end positions are not recovered in this case is because `PositionAt` truncates the values to [0,1]. `GetSegmentEnds` is not supposed to truncate the values. Values higher than 1 are desired because that indicates which segments are beyond the slider end.

Reply:
> It appears like you don't understand the output of the method.

I guess I don't, given that it's undocumented, and most of what I have to go on in terms of examples is this test which is misleading due to use of `PositionAt()` which truncates.

Reply:
> > It appears like you don't understand the output of the method.
> 
> I guess I don't, given that it's undocumented, and most of what I have to go on in terms of examples is this test which is misleading due to use of `PositionAt()` which truncates.

Yeah, I do think that the documentation needs a little addition, maybe something like this:
```diff
        /// <summary>
        /// Returns the progress values at which segments of the path end.
+       /// For segment ends beyond the path's <see cref="Distance"/>, values will be greater than 1.
        /// </summary>
        public IEnumerable<double> GetSegmentEnds()
```

As for the behaviour of the method, I think it's fine the way it is here. imo there's not really a point in truncating the values, it just gives whoever calls the method less information, and whether this is "expected behaviour" from a method called `GetSegmentEnds` just depends on who you ask, but the documentation should clarify that.

Reply:
I'll expand the documentation with the extra explanations I did in my previous reply. I hope that makes it 100% clear what the method is supposed to do.